File: java-ladidol/src/com/feng/newline/list/L92反转链表II.java
package com.feng.newline.list;

/**
 * @author: Xiaoqiang-Ladidol
 * @date: 2023/8/21 19:51
 * @description: {}
 */
public class L92反转链表II {


    //参考题解：https://leetcode.cn/problems/reverse-linked-list-ii/solutions/2397023/javapython3cfan-zhuan-lian-biao-chai-fen-ofh0/
    class Solution {


        ListNode reversePre;
        ListNode reverseHead;
        ListNode reverseTail;
        ListNode reverseNext;

        /**
         * 需要明确四个节点：
         * reversePre：反转区间的前一个节点；
         * reverseHead：反转区间的头节点；
         * reverseTail：反转区间的尾节点；
         * reverseNext：反转区间的下一个节点；
         *
         * @param head
         * @param left
         * @param right
         * @return
         */
        public ListNode reverseBetween(ListNode head, int left, int right) {
            int count = 1;

            ListNode dummy = new ListNode(-1, head);//虚拟头结点，来进行头结点的删除选择操作
            reversePre = dummy;
            while (count < left) {
                //找到翻转区间前一个元素
                reversePre = reversePre.next;
                count++;
            }
//            System.out.println("head = " + reversePre.val);

            reverseHead = reversePre.next;
            ListNode next = null;
            ListNode cur = reverseHead;
            ListNode pre = null;
            while (count <= right) {
                next = cur.next;
                cur.next = pre;
                pre = cur;
                cur = next;
                count++;
            }
            reverseNext = next;
            reverseTail = pre;

//            System.out.println("reverseNext = " + (reverseNext == null ? null : reverseNext.val));
//            System.out.println("reverseTail = " + reverseTail.val);
//            System.out.println("reverseHead = " + reverseHead.val);

            //开始连接回原链表
            reversePre.next = reverseTail;
            reverseHead.next = reverseNext;

            Integer nihao = 1;


            return dummy.next;

        }

    }


}

File: golang-devildyw/middle/L92反转链表II.go
package main

func main() {

}

// 切割出需要反转的部分 在使用反转链表的方式进行反转 最后在拼接起来就行
func reverseBetween(head *ListNode, left int, right int) *ListNode {
	// 切断我们需要反转的区间的连表 在使用反转链表的方式 吧他们反转 最后在接起来
	// 需要记录四个点

	// 虚拟头节点
	dummyNode := &ListNode{
		Val:  -1,
		Next: head,
	}

	pre := dummyNode

	// 找到left-1为止
	for i := 0; i < left-1; i++ {
		pre = pre.Next
	}

	// 找到right节点的位置
	rightNode := pre
	for i := 0; i < right-left+1; i++ {
		rightNode = rightNode.Next
	}

	// 切断出需要反转的链表 找到leftNode rightNode 已经找到了 保存一下末尾 一会儿需要拼接
	leftNode := pre.Next
	curr := rightNode.Next
	// 切断
	pre.Next = nil
	rightNode.Next = nil

	reverseLinkedList(leftNode)

	// 拼接
	pre.Next = rightNode
	leftNode.Next = curr
	return dummyNode.Next
}

func reverseLinkedList(head *ListNode) *ListNode {
	var tail *ListNode
	for head != nil {
		next := head.Next
		head.Next = tail
		tail = head
		head = next
	}
	return tail
}
